pkg/ingester/client/buffering_client.go

```diff
@@ -0,0 +1,111 @@
+package client
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/util/pool"
+)
+
+// This is a copy of (*ingesterClient).Push method, but accepting any message type.
+func pushRaw(ctx context.Context, conn *grpc.ClientConn, msg interface{}, opts ...grpc.CallOption) (*mimirpb.WriteResponse, error) {
+	out := new(mimirpb.WriteResponse)
+	err := conn.Invoke(ctx, "/cortex.Ingester/Push", msg, out, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+// bufferPoolingIngesterClient implements IngesterClient, but overrides Push method to add pooling of buffers used to marshal write requests.
+type bufferPoolingIngesterClient struct {
+	IngesterClient
+
+	conn *grpc.ClientConn
+
+	// This refers to pushRaw function, but is overridden in the benchmark to avoid doing actual grpc calls.
+	pushRawFn func(ctx context.Context, conn *grpc.ClientConn, msg interface{}, opts ...grpc.CallOption) (*mimirpb.WriteResponse, error)
+}
+
+func newBufferPoolingIngesterClient(client IngesterClient, conn *grpc.ClientConn) *bufferPoolingIngesterClient {
+	c := &bufferPoolingIngesterClient{
+		IngesterClient: client,
+		conn:           conn,
+		pushRawFn:      pushRaw,
+	}
+	return c
+}
+
+// Push wraps WriteRequest to implement buffer pooling.
+func (c *bufferPoolingIngesterClient) Push(ctx context.Context, in *mimirpb.WriteRequest, opts ...grpc.CallOption) (*mimirpb.WriteResponse, error) {
+	p := getPool(ctx)
+	if p == nil {
+		return c.IngesterClient.Push(ctx, in, opts...)
+	}
+
+	wr := &wrappedRequest{
+		WriteRequest: in,
+		slabPool:     p,
+	}
+	// We can return all buffers back to slabPool when this method finishes.
+	defer wr.ReturnBuffersToPool()
```

Comment:
The hard question is: is this safe or could the message buffer be retained by gRPC for longer? Can you share more details about the arguments on "this is safe to do". I still struggle a lot about this, because each time one of us look at it comes to a slightly different conclusion.

My question is not just about gRPC itself (for which you have added a nice test) but the chain of middlewares we have. Our middlewares shouldn't retain it, but logging may be async so worth to double check it.

Reply:
One thing to realize is that we're doing this buffering on the client side. "gRPC" that we talk about here is grpc connection to ingester.

Re middlewares... We setup our grpc connection with following interceptors:

- [grpcclient.Instrument](https://github.com/grafana/mimir/blob/c8d7c56ed225c706da534e230ff1d6892080db40/pkg/ingester/client/client.go#L42), which in turns [adds](https://github.com/grafana/mimir/blob/b5e7bce69311da4df36e363dce3c3d27c8ed0fb1/vendor/github.com/grafana/dskit/grpcclient/instrumentation.go#L13-L15):
  - `otgrpc.OpenTracingClientInterceptor(opentracing.GlobalTracer())`
  - `middleware.ClientUserHeaderInterceptor`
  - `middleware.UnaryClientInstrumentInterceptor(requestDuration)`
- Interceptors added by [`(*Config).DialOption`](https://github.com/grafana/mimir/blob/36f96f19bde5d6e5552600b25893244971b007c8/vendor/github.com/grafana/dskit/grpcclient/grpcclient.go#L95-L152): 
  - [BackoffRetry](https://github.com/grafana/mimir/blob/4b8cf2019a88d39231ab0b230701eeeb02079d12/vendor/github.com/grafana/dskit/grpcclient/backoff_retry.go#L14-L31)
  - [RateLimiter](https://github.com/grafana/mimir/blob/4b8cf2019a88d39231ab0b230701eeeb02079d12/vendor/github.com/grafana/dskit/grpcclient/ratelimit.go#L13-L26)

Looking at implementation of these interceptors...

- [`OpenTracingClientInterceptor`](https://github.com/grafana/mimir/blob/59fed803141e0a8dfbff32ea93cc6c4a7691ca4f/vendor/github.com/opentracing-contrib/go-grpc/client.go#L30-L75) has ability to 1) pass request to span inclusion function,, 2) log the payload (request) and 3) decorate the span using the request. Note that [we don't configure](https://github.com/grafana/mimir/blob/b5e7bce69311da4df36e363dce3c3d27c8ed0fb1/vendor/github.com/grafana/dskit/grpcclient/instrumentation.go#L13) `OpenTracingClientInterceptor` with any of these options.
- `middleware.ClientUserHeaderInterceptor`, `middleware.UnaryClientInstrumentInterceptor`, `BackoffRetry` and `RateLimiter` interceptors don't use request in any way (apart from passing it further)

Reply:
More places where our buffer could be passed around:
- [statsHandlers](https://github.com/grafana/mimir/blob/36f96f19bde5d6e5552600b25893244971b007c8/vendor/google.golang.org/grpc/stream.go#L1041-L1043)
- [binglogs](https://github.com/grafana/mimir/blob/36f96f19bde5d6e5552600b25893244971b007c8/vendor/google.golang.org/grpc/stream.go#L887-L895)

Let's start with binlogs. This is functionality configured by env variable `GRPC_BINARY_LOG_FILTER`. When configured, gRPC logs each RPC in the format defined by proto GrpcLogEntry. Design for this feature is at https://github.com/grpc/proposal/blob/master/A16-binary-logging.md.

When used, request is passed to logging method inside `binarylog.ClientMessage` struct. This struct has [`toProto()`](https://github.com/grafana/mimir/blob/36f96f19bde5d6e5552600b25893244971b007c8/vendor/google.golang.org/grpc/internal/binarylog/method_logger.go#L227) which will call another `Marhal` on our message. Once `binary.ClientMessage` is converted to `binlogpb.GrpcLogEntry` (now it contains our buffer from the pool), it's forwarded to the "sink" for writing.

There are three `sink` implementations in grpc: [`bufferedSink`](https://github.com/grafana/mimir/blob/f3afcb7a598c2147980371a64d985d20819fb49c/vendor/google.golang.org/grpc/internal/binarylog/sink.go#L99-L111), `noopSink` and [`writerSink`](https://github.com/grafana/mimir/blob/f3afcb7a598c2147980371a64d985d20819fb49c/vendor/google.golang.org/grpc/internal/binarylog/sink.go#L69-L84).

`writerSink` marshals `*binlogpb.GrpcLogEntry` and writes it to some writer. `bufferedSink` simply wraps `writerSink` but gives it a buffer to write output to, and also runs a goroutine to periodically flush the buffer. Important point is that it doesn't keep `*binlogpb.GrpcLogEntry` around, but marshals it immediately.

My conclusion is: using binary logging feature is safe when using our message buffering.

Reply:
Thanks for this analysis Peter. This is too valuable to get it lost in a GitHub comment. WDYT to move this to a `docs/contributing/...` page, where we can keep contributing the more we learn about it in the future?

Reply:
I've copied the notes above into doc: https://github.com/grafana/mimir/pull/5216

Reply:
Above analysis considered "happy path" only -- message gets sent to server, server replies. However in case of errors, messages can still be queued locally and sent later. https://github.com/grafana/mimir/pull/5830 fixes this.